File: sdk-go/preview.go
package boxkite

import (
	"context"
	"fmt"
	"net/url"
)

// PreviewURL is the response shape from CreatePreviewURL.
type PreviewURL struct {
	URL       string `json:"url"`
	ExpiresAt string `json:"expires_at"`
	TokenID   string `json:"token_id"`
}

// CreatePreviewURL mints a signed, time-limited URL that proxies HTTP
// traffic to a port a background process opened inside this session (see
// StartProcessOptions.ExposePort) (POST /v1/sandboxes/{id}/preview/{port}).
// The returned URL carries its own authorization -- no API key is required
// to use it, only to mint it.
//
// ttlSeconds bounds how long the minted URL stays valid (30-86400).
// Defaults to 900 (15 minutes) server-side when nil.
func (c *Client) CreatePreviewURL(ctx context.Context, sessionID string, port int, ttlSeconds *int) (*PreviewURL, error) {
	body := map[string]any{}
	if ttlSeconds != nil {
		body["ttl_seconds"] = *ttlSeconds
	}
	var out PreviewURL
	path := fmt.Sprintf("/v1/sandboxes/%s/preview/%d", url.PathEscape(sessionID), port)
	if err := c.doJSON(ctx, "POST", path, body, &out, nil); err != nil {
		return nil, err
	}
	return &out, nil
}

// PreviewRevokeResult is the response shape from RevokePreviewURL.
type PreviewRevokeResult struct {
	Revoked bool   `json:"revoked"`
	TokenID string `json:"token_id"`
}

// RevokePreviewURL invalidates one specific preview-URL token (its
// TokenID from CreatePreviewURL) before its TTL expires, without tearing
// down the sandbox session and without affecting any other preview token
// minted for the same session/port
// (POST /v1/sandboxes/{id}/preview/{port}/revoke). Idempotent: revoking
// an already-revoked, already-expired, or unrecognized tokenID still
// returns Revoked=true rather than erroring.
func (c *Client) RevokePreviewURL(ctx context.Context, sessionID string, port int, tokenID string) (*PreviewRevokeResult, error) {
	body := map[string]string{"token_id": tokenID}
	var out PreviewRevokeResult
	path := fmt.Sprintf("/v1/sandboxes/%s/preview/%d/revoke", url.PathEscape(sessionID), port)
	if err := c.doJSON(ctx, "POST", path, body, &out, nil); err != nil {
		return nil, err
	}
	return &out, nil
}

File: sdk-rust/src/preview.rs
//! Network-ingress preview URLs (`docs/NETWORK-INGRESS-DESIGN.md`):
//! `POST /v1/sandboxes/{id}/preview/{port}` and its `/revoke` sibling.
//! Mirrors `sdk-python`'s `create_preview_url`/`revoke_preview_url` and
//! `sdk-go`'s `CreatePreviewURL`/`RevokePreviewURL`.

use reqwest::Method;
use serde::{Deserialize, Serialize};

use crate::client::Client;
use crate::error::BoxkiteError;

/// A signed, time-limited preview URL (`POST /v1/sandboxes/{id}/preview/{port}`'s
/// response). The URL carries its own authorization -- no API key is needed
/// to use it, only to mint it.
#[derive(Debug, Clone, Deserialize)]
pub struct PreviewUrl {
    pub url: String,
    pub expires_at: String,
    pub token_id: String,
}

/// `POST /v1/sandboxes/{id}/preview/{port}/revoke`'s response.
#[derive(Debug, Clone, Deserialize)]
pub struct PreviewRevokeResult {
    pub revoked: bool,
    pub token_id: String,
}

impl Client {
    /// `POST /v1/sandboxes/{session_id}/preview/{port}` -- mint a signed,
    /// time-limited URL that proxies HTTP traffic to a port a background
    /// process opened inside this session (see
    /// [`StartProcessOptions::expose_port`](crate::StartProcessOptions::expose_port)).
    ///
    /// `ttl_seconds` bounds how long the minted URL stays valid (30-86400);
    /// defaults to 900 (15 minutes) server-side when `None`.
    pub async fn create_preview_url(
        &self,
        session_id: &str,
        port: u16,
        ttl_seconds: Option<u32>,
    ) -> Result<PreviewUrl, BoxkiteError> {
        #[derive(Serialize)]
        struct Body {
            #[serde(skip_serializing_if = "Option::is_none")]
            ttl_seconds: Option<u32>,
        }
        let builder = self
            .request(
                Method::POST,
                &format!("/v1/sandboxes/{session_id}/preview/{port}"),
            )
            .json(&Body { ttl_seconds });
        self.send(builder).await
    }

    /// `POST /v1/sandboxes/{session_id}/preview/{port}/revoke` -- invalidate
    /// one specific preview-URL token (its `token_id` from
    /// [`Client::create_preview_url`]) before its TTL expires, without
    /// tearing down the session and without affecting any other token minted
    /// for the same session/port. Idempotent: revoking an already-revoked,
    /// already-expired, or unrecognized `token_id` still returns
    /// `revoked = true` rather than erroring.
    pub async fn revoke_preview_url(
        &self,
        session_id: &str,
        port: u16,
        token_id: &str,
    ) -> Result<PreviewRevokeResult, BoxkiteError> {
        #[derive(Serialize)]
        struct Body<'a> {
            token_id: &'a str,
        }
        let builder = self
            .request(
                Method::POST,
                &format!("/v1/sandboxes/{session_id}/preview/{port}/revoke"),
            )
            .json(&Body { token_id });
        self.send(builder).await
    }
}
